security-utils/src/main/java/com/yahoo/security/tls/ReloadingTlsContext.java

```diff
@@ -34,19 +50,77 @@
 
     public ReloadingTlsContext(Path tlsOptionsConfigFile, AuthorizationMode mode) {
         this.tlsOptionsConfigFile = tlsOptionsConfigFile;
-        this.mode = mode;
-        this.currentTlsContext = new AtomicReference<>(new DefaultTlsContext(tlsOptionsConfigFile, mode));
-        this.scheduler.scheduleAtFixedRate(new SslContextReloader(),
+        TransportSecurityOptions options = TransportSecurityOptions.fromJsonFile(tlsOptionsConfigFile);
+        reloadCryptoMaterial(options, trustManager, keyManager);
+        this.tlsContext = createDefaultTlsContext(options, mode, trustManager, keyManager);
+        this.scheduler.scheduleAtFixedRate(new CryptoMaterialReloader(),
                                            UPDATE_PERIOD.getSeconds()/*initial delay*/,
                                            UPDATE_PERIOD.getSeconds(),
                                            TimeUnit.SECONDS);
     }
 
-    @Override
-    public SSLEngine createSslEngine() {
-        return currentTlsContext.get().createSslEngine();
+    private static void reloadCryptoMaterial(TransportSecurityOptions options,
+                                             MutableX509TrustManager trustManager,
+                                             MutableX509KeyManager keyManager) {
+        if (options.getCaCertificatesFile().isPresent()) {
+            trustManager.updateTruststore(loadTruststore(options.getCaCertificatesFile().get()));
+        } else {
+            trustManager.useDefaultTruststore();
```

Comment:
If a TLS config file has been explicitly provided, is there a useful case for using a default trust store? IIRC the C++ implementation throws if there is no CA file specified.

Reply:
The cacerts file from OpenJDK is pretty extensive and includes CAs such as _Let's Encrypt_.

I agree that it's not really useful here since the C++ implementation requires a CA file.